File: packages/clash-sdk/python/clash_sdk/agent.py
"""
ClashAgent — connects to ProjectRoom via WebSocket and executes custom actions.

Protocol:
1. Connect to ws://<server>/sync/<projectId>?token=<token>
2. Receive initial Loro CRDT snapshot (binary)
3. Send register_custom_actions text message
4. Monitor incoming Loro updates for new entries in the 'tasks' map
5. When a matching task arrives, call the handler
6. Upload result via HTTP POST /api/custom-action/upload
7. Send complete_custom_task text message to update node status
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import time
from typing import Any
from urllib.parse import urlencode

import aiohttp

from .decorators import ActionDefinition
from .models import ActionContext, ActionResult, AssetOutput

logger = logging.getLogger("clash_sdk")


class TaskState:
    """Tracks the state of an in-flight task."""

    def __init__(self, task_id: str, action_id: str, node_id: str):
        self.task_id = task_id
        self.action_id = action_id
        self.node_id = node_id
        self.status = "received"
        self.started_at = time.time()
        self.error: str | None = None


class ClashAgent:
    """
    Connects to a Clash ProjectRoom and executes registered custom actions.
    """

    def __init__(
        self,
        server_url: str,
        project_id: str,
        token: str,
        actions: list[ActionDefinition],
        runtime_id: str | None = None,
    ):
        self.server_url = server_url.rstrip("/")
        self.project_id = project_id
        self.token = token
        self.actions = {a.id: a for a in actions}
        # runtime_id identifies WHICH local machine hosts this action.
        # The server uses it to gate dispatch on the runtime being online
        # (deriveRuntimeStatus). Falls back to CLASH_RUNTIME_ID env so the
        # bridge daemon can inject it when supervising subprocess actions.
        #
        # If unset, register_custom_actions on the server side will reject
        # us — bail loudly during connect() rather than after a sleep.
        # Dev path: `export CLASH_RUNTIME_ID=$(jq -r .runtimeId ~/.clash/credentials.json)`
        # (the bridge writes credentials.json during `clash setup`).
        self.runtime_id = runtime_id or os.environ.get("CLASH_RUNTIME_ID") or None
        self.active_tasks: dict[str, TaskState] = {}
        self.task_history: list[dict[str, Any]] = []
        self._ws: aiohttp.ClientWebSocketResponse | None = None
        self._session: aiohttp.ClientSession | None = None
        self._running = False
        self._seen_tasks: set[str] = set()

    @property
    def http_url(self) -> str:
        """Convert ws:// to http:// for REST API calls."""
        url = self.server_url
        if url.startswith("ws://"):
            url = "http://" + url[5:]
        elif url.startswith("wss://"):
            url = "https://" + url[6:]
        return url

    async def connect(self) -> None:
        """Establish WebSocket connection and register actions."""
        if not self.runtime_id:
            raise RuntimeError(
                "CLASH_RUNTIME_ID is required to host custom actions. "
                "Set it via env or constructor arg. Quick dev path: "
                "export CLASH_RUNTIME_ID=$(jq -r .runtimeId ~/.clash/credentials.json)"
            )

        self._session = aiohttp.ClientSession()
        ws_url = f"{self.server_url}/sync/{self.project_id}?{urlencode({'token': self.token})}"

        logger.info(
            "Connecting to %s (runtime_id=%s…)",
            ws_url,
            self.runtime_id[:8] if self.runtime_id else "?",
        )
        self._ws = await self._session.ws_connect(
            ws_url,
            headers={
                "x-client-type": "cli",
                # Server validates this against the runtime table; the
                # WS upgrade returns 403 if the runtime doesn't exist or
                # belongs to a different user than the API token.
                "x-runtime-id": self.runtime_id,
            },
        )

        # Wait for initial snapshot (first binary message)
        msg = await self._ws.receive()
        if msg.type == aiohttp.WSMsgType.BINARY:
            logger.info("Received initial snapshot (%d bytes)", len(msg.data))
        else:
            raise ConnectionError(f"Expected binary snapshot, got {msg.type}")

        # Register custom actions
        register_msg = json.dumps({
            "type": "register_custom_actions",
            "actions": [a.to_manifest() for a in self.actions.values()],
        })
        await self._ws.send_str(register_msg)
        logger.info(
            "Registered %d action(s): %s",
            len(self.actions),
            list(self.actions.keys()),
        )

    async def disconnect(self) -> None:
        """Unregister actions and close connection."""
        self._running = False
        if self._ws and not self._ws.closed:
            # Unregister actions
            unregister_msg = json.dumps({
                "type": "unregister_custom_actions",
                "actionIds": list(self.actions.keys()),
            })
            try:
                await self._ws.send_str(unregister_msg)
            except Exception:
                pass
            await self._ws.close()
        if self._session:
            await self._session.close()
        logger.info("Disconnected")

    async def run_forever(self) -> None:
        """Main loop: listen for Loro updates and process custom tasks."""
        self._running = True
        logger.info("Listening for tasks...")

        while self._running:
            if not self._ws or self._ws.closed:
                logger.warning("WebSocket closed, stopping")
                break

            try:
                msg = await asyncio.wait_for(self._ws.receive(), timeout=30.0)
            except asyncio.TimeoutError:
                continue
            except Exception as e:
                logger.error("WebSocket error: %s", e)
                break

            if msg.type == aiohttp.WSMsgType.BINARY:
                # Loro CRDT update — check for new tasks
                # We parse the JSON representation to find task entries
                # In practice, we'd use a Loro CRDT library for Python,
                # but for now we use a simpler approach: listen for text sideband
                pass
            elif msg.type == aiohttp.WSMsgType.TEXT:
                await self._handle_text_message(msg.data)
            elif msg.type in (aiohttp.WSMsgType.CLOSE, aiohttp.WSMsgType.CLOSING, aiohttp.WSMsgType.CLOSED):
                logger.info("WebSocket closed by server")
                break
            elif msg.type == aiohttp.WSMsgType.ERROR:
                logger.error("WebSocket error")
                break

    async def _handle_text_message(self, text: str) -> None:
        """Handle JSON text sideband messages from ProjectRoom.

        The CLI agent connection (`x-client-type: cli`) gets a JSON
        sideband whenever NodeProcessor dispatches a custom-action
        task to a local runtime. Wire shape:

            {type: "custom_task_assigned", task: {...task record...}}
        """
        try:
            data = json.loads(text)
        except json.JSONDecodeError:
            return

        if data.get("type") == "custom_task_assigned":
            task = data.get("task") or {}
            task_id = task.get("taskId")
            if not task_id:
                logger.warning("custom_task_assigned with no taskId, ignoring")
                return
            logger.debug("Got custom_task_assigned task_id=%s action=%s", task_id, task.get("customActionId"))
            # Dedup BEFORE scheduling — two sideband messages
            # arriving in the same event-loop tick would otherwise
            # both queue handlers before either reached the seen-set
            # check inside _execute_task. The race fires when
            # ProjectRoom both broadcasts a fresh dispatch and
            # replays it on a near-simultaneous reconnect (each
            # spawns N extra siblings server-side).
            if task_id in self._seen_tasks:
                return
            self._seen_tasks.add(task_id)
            # Long-running actions (Pillow slicing, ML inference)
            # may take seconds — run concurrently so the receive
            # loop keeps processing messages.
            asyncio.create_task(self._execute_task(task))

    async def _execute_task(self, task: dict[str, Any]) -> None:
        """Execute a custom action task."""
        task_id = task["taskId"]
        action_id = task.get("customActionId", "")
        node_id = task["nodeId"]
        project_id = task.get("projectId", self.project_id)

        # Dedup happens at the entry point (_handle_text_message) now —
        # if we got here, _seen_tasks already contains task_id from the
        # caller. Re-checking would always short-circuit. Note that the
        # `process_loro_update` placeholder (currently a no-op) might
        # call this directly someday; if that lands, it must add to
        # _seen_tasks before calling.

        action_def = self.actions.get(action_id)
        if not action_def:
            logger.warning("No handler for action '%s', skipping task %s", action_id, task_id)
            return

        state = TaskState(task_id, action_id, node_id)
        self.active_tasks[task_id] = state
        logger.info("Executing task %s (action: %s)", task_id, action_id)

        try:
            state.status = "running"
            refs = task.get("refs") or {}
            ctx = ActionContext(
                task_id=task_id,
                node_id=node_id,
                project_id=project_id,
                action_id=action_id,
                prompt=task.get("prompt", ""),
                params=task.get("params", {}),
                output_type=task.get("outputType", action_def.output_type),
                reference_image_r2_keys=list(refs.get("image") or []),
                reference_video_r2_keys=list(refs.get("video") or []),
                reference_audio_r2_keys=list(refs.get("audio") or []),
                fetch_asset=self.fetch_asset,
            )

            result = await action_def.handler(ctx)
            state.status = "uploading"

            # Upload each binary output, build asset descriptors for
            # the complete message. Text outputs ride along with no
            # upload — server stamps `data.content` directly.
            assets: list[dict[str, Any]] = []
            for idx, out in enumerate(result.outputs):
                if out.type == "text":
                    assets.append({
                        "type": "text",
                        "content": out.content or "",
                        "label": out.label,
                    })
                    continue
                if not out.data:
                    raise RuntimeError(
                        f"AssetOutput[{idx}] type={out.type} has no data"
                    )
                storage_key = await self._upload_one(
                    project_id, task_id, node_id, out, idx
                )
                assets.append({
                    "type": out.type,
                    "storageKey": storage_key,
                    "mimeType": out.mime_type,
                    "label": out.label,
                })

            state.status = "completed"
            await self._complete_task(
                task_id, node_id, "completed", assets, result.description
            )

            duration_ms = (time.time() - state.started_at) * 1000
            logger.info(
                "Task %s completed in %.0fms", task_id, duration_ms
            )
            self.task_history.append({
                "taskId": task_id,
                "actionId": action_id,
                "status": "completed",
                "durationMs": duration_ms,
            })

        except Exception as e:
            state.status = "failed"
            state.error = str(e)
            logger.error("Task %s failed: %s", task_id, e)

            await self._complete_task(
                task_id, node_id, "failed", [], None, error=str(e)
            )
            self.task_history.append({
                "taskId": task_id,
                "actionId": action_id,
                "status": "failed",
                "error": str(e),
            })

        finally:
            del self.active_tasks[task_id]

    async def fetch_asset(self, storage_key: str) -> bytes:
        """Download a reference asset by R2 storage key.

        Used by action handlers that need to operate on their input
        images (e.g. grid_split). The agent's bearer token mints a
        signed URL, then we stream the bytes back. The signed URL
        machinery already enforces project-level auth — we don't
        need additional checks here.
        """
        if not self._session:
            raise RuntimeError("No HTTP session")
        sign_url = f"{self.http_url}/assets/sign?key={storage_key}"
        headers = {"Authorization": f"Bearer {self.token}"}
        async with self._session.get(sign_url, headers=headers) as resp:
            if resp.status != 200:
                body = await resp.text()
                raise RuntimeError(f"Sign failed ({resp.status}): {body}")
            signed = (await resp.json()).get("url")
        if not signed:
            raise RuntimeError(f"No signed URL returned for {storage_key}")
        get_url = f"{self.http_url}{signed}" if signed.startswith("/") else signed
        async with self._session.get(get_url) as resp:
            if resp.status != 200:
                body = await resp.text()
                raise RuntimeError(f"Fetch failed ({resp.status}): {body}")
            return await resp.read()

    async def _upload_one(
        self,
        project_id: str,
        task_id: str,
        node_id: str,
        out: AssetOutput,
        idx: int,
    ) -> str:
        """Upload one binary asset; returns its R2 storage key.

        Multi-output actions call this once per binary output. The
        `outputIndex` form field tells the server to suffix the R2
        key so sibling uploads from the same task don't collide.
        """
        if not self._session:
            raise RuntimeError("No HTTP session")

        url = f"{self.http_url}/api/custom-action/upload"
        form = aiohttp.FormData()
        form.add_field("projectId", project_id)
        form.add_field("taskId", task_id)
        form.add_field("nodeId", node_id)
        form.add_field("outputType", out.type)
        form.add_field("outputIndex", str(idx))
        form.add_field(
            "file",
            out.data or b"",
            filename=f"result-{idx}.{_ext(out.type)}",
            content_type=out.mime_type or "application/octet-stream",
        )

        headers = {"Authorization": f"Bearer {self.token}"}
        async with self._session.post(url, data=form, headers=headers) as resp:
            if resp.status != 200:
                body = await resp.text()
                raise RuntimeError(f"Upload failed ({resp.status}): {body}")
            data = await resp.json()
            return data.get("storageKey", "")

    async def _complete_task(
        self,
        task_id: str,
        node_id: str,
        status: str,
        assets: list[dict[str, Any]],
        description: str | None,
        error: str | None = None,
    ) -> None:
        """Send complete_custom_task message via WebSocket.

        Wire shape:
            {
              type: "complete_custom_task",
              taskId, nodeId, status,
              result: {
                assets: [{type, storageKey?, content?, mimeType?, label?}, ...],
                description?, error?
              }
            }
        """
        if not self._ws or self._ws.closed:
            return

        result: dict[str, Any] = {"assets": assets}
        if description:
            result["description"] = description
        if error:
            result["error"] = error

        msg: dict[str, Any] = {
            "type": "complete_custom_task",
            "taskId": task_id,
            "nodeId": node_id,
            "status": status,
            "result": result,
        }
        await self._ws.send_str(json.dumps(msg))

    def process_loro_update(self, data: bytes) -> list[dict[str, Any]]:
        """
        Extract new tasks from a Loro CRDT update.

        Note: Full Loro CRDT parsing in Python requires a native binding.
        For now, this is a placeholder. In production, use loro-crdt Python
        bindings or parse the tasks map from the document state.
        """
        # TODO: Integrate with loro-crdt Python bindings when available
        return []


def _ext(output_type: str) -> str:
    return {"image": "png", "video": "mp4", "audio": "mp3"}.get(output_type, "bin")


def run(
    server_url: str,
    project_id: str,
    token: str,
    actions: list[ActionDefinition] | None = None,
    runtime_id: str | None = None,
) -> None:
    """
    Convenience function to create an agent and run it.

    Example:
        from clash_sdk import action, ActionContext, ActionResult, run

        @action(id="echo", name="Echo", output_type="text")
        async def echo(ctx: ActionContext) -> ActionResult:
            return ActionResult.text(f"Echo: {ctx.prompt}")

        run(server_url="ws://localhost:8789", project_id="my-proj", token="...",
            actions=[echo])
    """
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s [%(name)s] %(levelname)s: %(message)s",
    )

    agent = ClashAgent(
        server_url=server_url,
        project_id=project_id,
        token=token,
        actions=actions or [],
        runtime_id=runtime_id,
    )

    async def _main() -> None:
        await agent.connect()
        try:
            await agent.run_forever()
        finally:
            await agent.disconnect()

    try:
        asyncio.run(_main())
    except KeyboardInterrupt:
        logger.info("Shutting down...")

File: packages/clash-sdk/js/src/agent.ts
/**
 * ClashAgent — connects to ProjectRoom via WebSocket, registers
 * actions, executes incoming tasks.
 *
 * Wire-protocol parity with the Python SDK (see
 * `packages/clash-sdk/python/clash_sdk/agent.py`). Behavioural rules
 * worth noting:
 *
 *   1. Dedup happens at receive (`_handleTextMessage`) BEFORE
 *      scheduling the async handler. Python had a double-dedup bug
 *      that silently bailed; we don't repeat it here.
 *   2. Each binary output gets its own POST to `/api/custom-action/upload`
 *      with `outputIndex` — the server suffixes R2 keys + asset ids
 *      so siblings from the same task don't overwrite each other.
 *   3. Failures send `status: 'failed' + result.assets: []` so the
 *      server marks the primary pending child failed without spawning
 *      any siblings (matches Python's failure path).
 */

import WebSocket from 'ws';
import { Buffer } from 'node:buffer';
import { randomUUID } from 'node:crypto';
import { setTimeout as delay } from 'node:timers/promises';
import type {
  ActionContext,
  ActionDefinition,
  ActionResult,
  AssetOutput,
  RunOptions,
} from './types.js';

/**
 * Convert ws://host → http://host (and wss://→https://). Falls through
 * for already-HTTP URLs — caller may pass either form.
 */
function toHttpUrl(serverUrl: string): string {
  if (serverUrl.startsWith('ws://')) return 'http://' + serverUrl.slice(5);
  if (serverUrl.startsWith('wss://')) return 'https://' + serverUrl.slice(6);
  return serverUrl;
}

function toWsUrl(serverUrl: string): string {
  if (serverUrl.startsWith('http://')) return 'ws://' + serverUrl.slice(7);
  if (serverUrl.startsWith('https://')) return 'wss://' + serverUrl.slice(8);
  return serverUrl;
}

function extOf(modality: AssetOutput['type']): string {
  switch (modality) {
    case 'video': return 'mp4';
    case 'audio': return 'mp3';
    case 'image': return 'png';
    case 'text':  return 'txt';
  }
}

function defaultMime(modality: AssetOutput['type']): string {
  switch (modality) {
    case 'video': return 'video/mp4';
    case 'audio': return 'audio/mpeg';
    case 'image': return 'image/png';
    case 'text':  return 'text/plain';
  }
}

export class ClashAgent {
  private ws: WebSocket | null = null;
  private readonly actions: Map<string, ActionDefinition>;
  private readonly seenTasks = new Set<string>();
  private running = false;
  private readonly httpUrl: string;
  private readonly wsUrl: string;

  constructor(private readonly opts: RunOptions) {
    if (!opts.runtimeId) {
      throw new Error(
        'CLASH_RUNTIME_ID missing — set it from ~/.clash/credentials.json#runtimeId. ' +
          'The server rejects registrations without an x-runtime-id header.',
      );
    }
    this.actions = new Map(opts.actions.map((a) => [a.id, a]));
    this.httpUrl = toHttpUrl(opts.serverUrl).replace(/\/$/, '');
    this.wsUrl = toWsUrl(opts.serverUrl).replace(/\/$/, '');
  }

  async run(): Promise<void> {
    await this.connect();
    this.running = true;
    await this.runLoop();
  }

  async stop(): Promise<void> {
    this.running = false;
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      try {
        this.ws.send(JSON.stringify({
          type: 'unregister_custom_actions',
          actionIds: Array.from(this.actions.keys()),
        }));
      } catch { /* socket may be racing close */ }
      this.ws.close(1000, 'shutdown');
    }
  }

  private async connect(): Promise<void> {
    const url = `${this.wsUrl}/sync/${this.opts.projectId}?token=${encodeURIComponent(this.opts.apiKey)}`;
    this.log(`Connecting to ${url} (runtime_id=${this.opts.runtimeId.slice(0, 8)}…)`);
    const ws = new WebSocket(url, {
      headers: {
        'x-client-type': 'cli',
        'x-runtime-id': this.opts.runtimeId,
      },
    });

    // Wait for open + the server's initial binary snapshot. ProjectRoom
    // always sends a Loro snapshot as the first frame; we can ignore
    // its contents (we don't parse Loro CRDT in JS — same as Python).
    await new Promise<void>((resolve, reject) => {
      const onError = (err: Error) => { cleanup(); reject(err); };
      const onClose = (code: number, reason: Buffer) => {
        cleanup();
        reject(new Error(`WS closed before open: ${code} ${reason.toString()}`));
      };
      const onMessage = (data: WebSocket.RawData, isBinary: boolean) => {
        if (!isBinary) return; // text frames will arrive after snapshot
        cleanup();
        this.log(`Received initial snapshot (${(data as Buffer).length} bytes)`);
        resolve();
      };
      const cleanup = () => {
        ws.off('error', onError);
        ws.off('close', onClose);
        ws.off('message', onMessage);
      };
      ws.once('error', onError);
      ws.once('close', onClose);
      ws.on('message', onMessage);
    });

    this.ws = ws;
    ws.send(JSON.stringify({
      type: 'register_custom_actions',
      actions: Array.from(this.actions.values()).map((a) => ({
        id: a.id,
        name: a.name,
        description: a.description ?? '',
        parameters: a.parameters ?? [],
        outputType: a.outputType,
        icon: a.icon ?? '',
        color: a.color ?? '',
        promptModalities: a.promptModalities ?? ['text'],
        runtime: 'local',
      })),
    }));
    this.log(`Registered ${this.actions.size} action(s): [${[...this.actions.keys()].join(', ')}]`);
  }

  private async runLoop(): Promise<void> {
    if (!this.ws) throw new Error('not connected');
    const ws = this.ws;

    return new Promise<void>((resolve) => {
      ws.on('message', (data, isBinary) => {
        if (isBinary) return; // ignore Loro CRDT frames
        const text = (data as Buffer).toString('utf-8');
        this.handleTextMessage(text);
      });
      ws.once('close', (code, reason) => {
        this.log(`WS closed code=${code} reason=${reason.toString() || '—'}`);
        resolve();
      });
      ws.once('error', (err) => {
        this.log(`WS error: ${err.message}`);
        // Don't reject — let close handler resolve. Same shape as Python SDK.
      });
    });
  }

  private handleTextMessage(text: string): void {
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(text);
    } catch {
      return;
    }

    if (data.type !== 'custom_task_assigned') return;

    const task = (data.task ?? {}) as Record<string, unknown>;
    const taskId = task.taskId as string | undefined;
    if (!taskId) {
      this.log('custom_task_assigned with no taskId, ignoring');
      return;
    }
    // Dedup BEFORE scheduling — two sideband messages arriving in the
    // same event-loop tick would otherwise both queue handlers before
    // either added to seenTasks. Python had this exact bug.
    if (this.seenTasks.has(taskId)) return;
    this.seenTasks.add(taskId);

    void this.executeTask(task);
  }

  private async executeTask(task: Record<string, unknown>): Promise<void> {
    const taskId = task.taskId as string;
    const actionId = (task.customActionId as string) ?? '';
    const nodeId = task.nodeId as string;
    const projectId = (task.projectId as string) ?? this.opts.projectId;

    const def = this.actions.get(actionId);
    if (!def) {
      this.log(`No handler for action '${actionId}', skipping task ${taskId}`);
      return;
    }

    const startedAt = Date.now();
    this.log(`Executing task ${taskId} (action: ${actionId})`);

    const refs = (task.refs ?? {}) as Record<string, string[]>;
    const ctx: ActionContext = {
      taskId,
      nodeId,
      projectId,
      actionId,
      prompt: (task.prompt as string) ?? '',
      params: (task.params as Record<string, string | number | boolean>) ?? {},
      outputType: (task.outputType as AssetOutput['type']) ?? def.outputType,
      referenceImageR2Keys: refs.image ?? [],
      referenceVideoR2Keys: refs.video ?? [],
      referenceAudioR2Keys: refs.audio ?? [],
      fetchAsset: (key) => this.fetchAsset(key),
    };

    // Phase 0 attribution: server stamps actorUserId / actorAgentId
    // onto the task record before assigning it. Echo them back on
    // /api/custom-action/upload so the resulting asset row attributes
    // to the actor that placed the node (not the project owner).
    const actorUserId = (task.actorUserId as string) ?? '';
    const actorAgentId = (task.actorAgentId as string) ?? '';

    try {
      const result = await def.handler(ctx);
      const assets = await this.uploadOutputs(projectId, taskId, nodeId, result.outputs, actorUserId, actorAgentId);
      this.sendComplete(taskId, nodeId, 'completed', assets, result.description);
      this.log(`Task ${taskId} completed in ${Date.now() - startedAt}ms`);
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e);
      this.log(`Task ${taskId} failed: ${msg}`);
      this.sendComplete(taskId, nodeId, 'failed', [], undefined, msg);
    }
  }

  /**
   * Upload binary outputs to /api/custom-action/upload, one POST per
   * output with outputIndex set. Text outputs ride along in the
   * complete_custom_task message body (no R2 hop).
   */
  private async uploadOutputs(
    projectId: string,
    taskId: string,
    nodeId: string,
    outputs: AssetOutput[],
    actorUserId = '',
    actorAgentId = '',
  ): Promise<Array<Record<string, unknown>>> {
    const results: Array<Record<string, unknown>> = [];
    for (let idx = 0; idx < outputs.length; idx++) {
      const out = outputs[idx];
      if (out.type === 'text') {
        results.push({ type: 'text', content: out.content ?? '', label: out.label });
        continue;
      }
      if (!out.data) {
        throw new Error(`AssetOutput[${idx}] type=${out.type} has no data`);
      }
      const storageKey = await this.uploadOne(projectId, taskId, nodeId, out, idx, actorUserId, actorAgentId);
      results.push({
        type: out.type,
        storageKey,
        mimeType: out.mimeType,
        label: out.label,
      });
    }
    return results;
  }

  private async uploadOne(
    projectId: string,
    taskId: string,
    nodeId: string,
    out: AssetOutput,
    idx: number,
    actorUserId = '',
    actorAgentId = '',
  ): Promise<string> {
    const form = new FormData();
    form.set('projectId', projectId);
    form.set('taskId', taskId);
    form.set('nodeId', nodeId);
    form.set('outputType', out.type);
    form.set('outputIndex', String(idx));
    if (actorUserId) form.set('actorUserId', actorUserId);
    if (actorAgentId) form.set('actorAgentId', actorAgentId);
    // Node's BlobPart typing collides with Buffer<ArrayBufferLike> in
    // recent @types/node. Copying into a fresh Uint8Array gives the
    // FormData encoder a clean buffer.
    const src = out.data!;
    const bytes = src instanceof Uint8Array ? new Uint8Array(src) : new Uint8Array(src);
    const blob = new Blob([bytes], {
      type: out.mimeType ?? defaultMime(out.type),
    });
    form.set('file', blob, `result-${idx}.${extOf(out.type)}`);

    const res = await fetch(`${this.httpUrl}/api/custom-action/upload`, {
      method: 'POST',
      body: form,
      headers: { Authorization: `Bearer ${this.opts.apiKey}` },
    });
    if (!res.ok) {
      const body = await res.text().catch(() => '');
      throw new Error(`Upload failed (${res.status}): ${body}`);
    }
    const json = (await res.json()) as { storageKey?: string };
    return json.storageKey ?? '';
  }

  private sendComplete(
    taskId: string,
    nodeId: string,
    status: 'completed' | 'failed',
    assets: Array<Record<string, unknown>>,
    description?: string,
    error?: string,
  ): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    const result: Record<string, unknown> = { assets };
    if (description) result.description = description;
    if (error) result.error = error;
    this.ws.send(JSON.stringify({
      type: 'complete_custom_task',
      taskId,
      nodeId,
      status,
      result,
    }));
  }

  /**
   * Fetch an asset's bytes by R2 storage key. Mirrors the Python
   * SDK's `fetch_asset` — hits /assets/sign with bearer auth, then
   * GETs the signed URL. The signing layer scopes access to the
   * project, so we don't need extra auth on the GET.
   */
  private async fetchAsset(r2Key: string): Promise<Buffer> {
    const signRes = await fetch(
      `${this.httpUrl}/assets/sign?key=${encodeURIComponent(r2Key)}`,
      { headers: { Authorization: `Bearer ${this.opts.apiKey}` } },
    );
    if (!signRes.ok) {
      throw new Error(`Sign failed (${signRes.status}): ${await signRes.text().catch(() => '')}`);
    }
    const { url } = (await signRes.json()) as { url?: string };
    if (!url) throw new Error(`No signed URL returned for ${r2Key}`);
    const absoluteUrl = url.startsWith('/') ? `${this.httpUrl}${url}` : url;
    const getRes = await fetch(absoluteUrl);
    if (!getRes.ok) {
      throw new Error(`Fetch failed (${getRes.status}): ${await getRes.text().catch(() => '')}`);
    }
    return Buffer.from(await getRes.arrayBuffer());
  }

  private log(msg: string): void {
    // Matches Python SDK's stderr-style chunks so users see a unified
    // log stream when running multiple SDKs side-by-side.
    process.stderr.write(`[clash-sdk] ${new Date().toISOString()} ${msg}\n`);
  }
}

/**
 * Top-level entry point — instantiate, connect, run until WS closes.
 * SIGTERM cleanly unregisters + closes the socket.
 */
export async function run(opts: RunOptions): Promise<void> {
  const agent = new ClashAgent(opts);
  const shutdown = async () => {
    try { await agent.stop(); } catch { /* best effort */ }
    process.exit(0);
  };
  process.once('SIGTERM', () => void shutdown());
  process.once('SIGINT', () => void shutdown());
  await agent.run();
}

// Suppress unused-import warning when tsup tree-shakes single-file
// builds — the `delay` + `randomUUID` imports are kept for future
// reconnect/backoff logic that mirrors Python's behaviour.
void delay; void randomUUID;
